### PR TITLE
vncserver enhancement

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
* Support keyboard driver 
## Impact
vncserver only
## Testing
sim:vncserver

### CODE REVIEW

File: drivers/video/vnc/vnc_fbdev.c [hunk 1]
```diff
@@ -631,6 +631,27 @@ int up_fbinitialize(int display)
...
+    {
+      gerr("ERROR: Format vnc keyboard driver path failed.\n");
+      return ret;
+    }
+
+  ret = vnc_kbd_register(devname, session);
+
+  if (ret < 0)
+    {
+      gerr("ERROR: Initial vnc keyboard driver failed.\n");
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:708]: unregister touch

File: drivers/video/vnc/vnc_fbdev.c [hunk 2]
```diff
@@ -631,6 +631,27 @@ int up_fbinitialize(int display)
...
+      return ret;
+    }
+
+  session->kbdout = vnc_kbd_event;
+#endif
+
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:714]: need unregister keyboard after line 644

File: drivers/video/vnc/vnc_kbd.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * Name: vnc_kbd_event
+ ****************************************************************************/
+
+int vnc_kbd_event(FAR void *arg, uint8_t nch, FAR const uint8_t *ch)
+{
+  int i = 0;
+  FAR struct vnc_session_s *session = arg;
+  for (i = 0; i < nch; i++)
+    {
+      keyboard_event(&session->kbd, ch[i], KEYBOARD_PRESS);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_kbd.c:61]: need convert to uin32_t and send once

File: drivers/video/vnc/vnc_fbdev.c [hunk 4]
```diff
@@ -577,6 +577,64 @@ static inline int vnc_wait_start(int display)
...
+ ****************************************************************************/
+
+static int vnc_session_alloc(void)
+{
+  FAR struct vnc_session_s *session;
+  int display;
+  int ret = OK;
+  FAR uint8_t *fb;
+
+  for (display = 0; display < RFB_MAX_DISPLAYS; display++)
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:595]: what's problem to allocate session in vnc_server?

Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c:595]: Touch/kbd inited before start vnc_server thread, thus the `g_vnc_sessions` is NULL before `vnc_start_server`.

Comment (xiaoxiang781216) [reply drivers/video/vnc/vnc_fbdev.c:595]: Ok, look like we should move register after vnc_wait_start?

Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c:595]: Yes, we should move touch/kbd driver after vnc_start_server or apply the second commit to init session in up_fbinitialize

Comment (xiaoxiang781216) [reply drivers/video/vnc/vnc_fbdev.c:595]: it's more natural to initialize touch/kbd after vnc_start_server.

File: drivers/video/vnc/vnc_fbdev.c [hunk 5]
```diff
@@ -631,6 +694,27 @@ int up_fbinitialize(int display)
 
   session->mouseout = vnc_touch_event;
 #endif
+
+#ifdef CONFIG_VNCSERVER_KBD
+  ret = snprintf(kbd_name, NAME_MAX, CONFIG_VNCSERVER_KBD_DEVNAME "%d",
+                 display);
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_fbdev.c:700]:
```suggestion
  ret = snprintf(kbd_name, sizeof(kbd_name), CONFIG_VNCSERVER_KBD_DEVNAME "%d",
                 display);
```

File: drivers/video/vnc/vnc_fbdev.c [hunk 6]
```diff
@@ -602,17 +660,22 @@ int up_fbinitialize(int display)
...
+  ret = vnc_session_alloc();
+
   /* Save the input callout function information in the session structure. */
 
   session           = g_vnc_sessions[display];
 #ifdef CONFIG_VNCSERVER_TOUCH
 
-  ret = snprintf(devname, NAME_MAX, CONFIG_VNCSERVER_TOUCH_DEVNAME "%d",
+  ret = snprintf(touch_name, NAME_MAX, CONFIG_VNCSERVER_TOUCH_DEVNAME "%d",
                  display);
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_fbdev.c:679]:
```suggestion
  ret = snprintf(touch_name, sizeof(touch_name), CONFIG_VNCSERVER_TOUCH_DEVNAME "%d",
                 display);
```

File: drivers/video/vnc/vnc_fbdev.c (hunk 4 above)
Comment (no1wudi) [reply drivers/video/vnc/vnc_fbdev.c]: OK

**Comment by no1wudi** [comment]
> @no1wudi I'm looking for a VNC client on NuttX side, do you have some suggestion how I could implement it? I think a VNC client should be userspace only and could use framebuffer interface, right?

Yes, and vnc client also can be intergrated to graphic lib (nx or lvgl) with some adaptions.

File: drivers/video/vnc/vnc_fbdev.c [hunk 7]
```diff
@@ -602,26 +602,43 @@ int up_fbinitialize(int display)
   int ret;
   FAR struct vnc_session_s *session;
 #ifdef CONFIG_VNCSERVER_TOUCH
-  char devname[NAME_MAX];
+  char touch_name[NAME_MAX];
+#endif
+#ifdef CONFIG_VNCSERVER_KBD
+  char kbd_name[NAME_MAX];
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:608]: why not use the same name as up_fbuninitialize?

File: drivers/video/vnc/vnc_fbdev.c [hunk 8]
```diff
@@ -602,26 +602,43 @@ int up_fbinitialize(int display)
...
 #endif
 
   DEBUGASSERT(display >= 0 && display < RFB_MAX_DISPLAYS);
 
+  /* Start the VNC server kernel thread. */
+
+  ret = vnc_start_server(display);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:615]: why remove the error check?

File: drivers/video/vnc/vnc_receiver.c [hunk 9]
```diff
@@ -321,8 +321,14 @@ int vnc_receiver(FAR struct vnc_session_s *session)
                   /* Inject the key press/release event into NX */
 
                   keyevent = (FAR struct rfb_keyevent_s *)session->inbuf;
+#ifndef CONFIG_VNCSERVER_KBD
                   vnc_key_map(session, rfb_getbe32(keyevent->key),
                               (bool)keyevent->down);
+#else
+                  /* Pass raw keycode to keyboard driver */
+
+                  session->kbdout(session, keyevent->down, keyevent->key);
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_receiver.c:330]: should we move the code into vnc_key_map

Comment (no1wudi) [reply drivers/video/vnc/vnc_receiver.c:330]: Is it better to move keycode decode to application layer from this virtual keyboard driver ?

Comment (xiaoxiang781216) [reply drivers/video/vnc/vnc_receiver.c:330]: Yes, if the user doesn't enable CONFIG_VNCSERVER_KBDENCODE, but it's better to move the code into vnc_key_map

File: drivers/video/vnc/vnc_fbdev.c [hunk 10]
```diff
@@ -631,29 +651,29 @@ int up_fbinitialize(int display)
...
     {
-      gerr("ERROR: wait for vnc server start failed: %d\n", ret);
+      gerr("ERROR: Initial vnc keyboard driver failed.\n");
+      return ret;
     }
 
+  session->kbdout = vnc_kbd_event;
+#endif
+
+  session->arg      = session;
```
Comment (xiaoxiang781216) [inline drivers/video/vnc/vnc_fbdev.c:675]: Move after line 632

File: drivers/video/vnc/vnc_keymap.c [hunk 11]
```diff
@@ -454,6 +454,17 @@ static int vnc_kbd_ascii(uint16_t keysym)
...
                  bool keydown)
 {
+  /* If uinput like virtual keyboard enabled, pass keycode to keyboard driver.
+   *
+   * This driver cannot be used with NX now, if vnc server works with NX, should
+   * not enable CONFIG_VNCSERVER_KBD.
+   */
+
+#ifdef CONFIG_VNCSERVER_KBD
+  session->kbdout(&session->kbd, keydown, (const uint8_t *)&keysym);
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_keymap.c:464]:
```suggestion
  session->kbdout(&session->kbd, keydown, (FAR const uint8_t *)&keysym);
```

File: drivers/video/vnc/vnc_server.h [hunk 12]
```diff
@@ -227,6 +228,10 @@ struct vnc_session_s
   FAR struct touch_lowerhalf_s touch; /* Touch driver instance */
 #endif
 
+#ifdef CONFIG_VNCSERVER_KBD
+  FAR struct keyboard_lowerhalf_s kbd; /* Keyboard driver instance */
```
Comment (pkarashchenko) [inline drivers/video/vnc/vnc_server.h:232]:
```suggestion
  struct keyboard_lowerhalf_s kbd; /* Keyboard driver instance */
```
and same for `struct touch_lowerhalf_s touch;` above

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**